Add source IP and port fields to Goldmane flow logs

ravikalla opened the pull request:
## Summary

This PR addresses issue #10601 by adding source IP address and port information to the Goldmane flows API. This enhancement provides security teams with valuable visibility into the source of network traffic, which is particularly useful for identifying malicious inbound traffic from external networks.

## Changes

### Core Implementation
- **Protobuf Schema**: Added `source_ip` (string) and `source_port` (int64) fields to the `FlowKey` message
- **Goldmane Types**: Updated Go structs to include source IP and port fields with appropriate getter methods
- **Felix Integration**: Modified flow conversion logic to extract IP addresses from network tuples and convert to string format
- **API Response**: Enhanced Whisker backend to include source IP/port in JSON responses

### Key Features
- **IPv4 and IPv6 Support**: Handles both address families correctly
- **Backward Compatibility**: Existing flows without source IP work seamlessly (empty strings/zero values)
- **Complete Data Flow**: Source IP flows from Felix → Goldmane → Whisker API
- **Efficient Format**: IP addresses stored as strings for easy consumption by UI and clients

## Testing

### Comprehensive Test Coverage
- **Unit Tests**: Added tests for all conversion functions and type transformations
- **Integration Tests**: End-to-end flow testing with various scenarios
- **Backward Compatibility**: Verified old flows continue to work unchanged  
- **Protobuf Compatibility**: Ensured old proto messages can be read by new code
- **API Compatibility**: Validated JSON schema remains compatible

### Test Scenarios Verified
- ✅ Pod-to-pod communication with IPv4/IPv6
- ✅ External traffic from public internet
- ✅ Legacy flows without IP information  
- ✅ Host endpoint traffic
- ✅ All endpoint types (WorkloadEndpoint, HostEndpoint, NetworkSet, Network)
- ✅ Edge cases (empty IPs, special addresses)

## Backward Compatibility

This PR maintains 100% backward compatibility:
- Existing API clients continue to work without changes
- Old flow data is processed correctly with empty source IP fields
- No breaking changes to protobuf schema or JSON API
- Wire format efficiency preserved

## Example

### Before (existing flows)
```json
{
  "source_name": "frontend-app",
  "source_namespace": "production", 
  "dest_name": "api-service",
  "dest_port": 443
}
```

### After (new flows with source IP)
```json
{
  "source_name": "frontend-app",
  "source_namespace": "production",
  "source_ip": "10.244.1.100",
  "source_port": 45678,
  "dest_name": "api-service", 
  "dest_port": 443
}
```

## Test Plan

1. **Build Verification**: All components (Goldmane, Felix, Whisker) build successfully
2. **Unit Tests**: Run existing and new test suites
3. **Integration Testing**: Deploy in test environment and verify flow data collection
4. **Performance Testing**: Validate no significant overhead introduced

Closes #10601

Commit:
Add source IP and port fields to Goldmane flow logs

Addresses issue #10601 by adding source IP address and port information
to the Goldmane flows API. This allows security teams to identify the
source of network traffic, which is particularly useful for detecting
malicious inbound traffic from external networks.

Changes:
- Add source_ip and source_port fields to FlowKey protobuf definition
- Update Goldmane Go types to include source IP/port fields
- Modify Felix conversion logic to extract IP data from flow tuples
- Update Whisker API response structures to include source IP/port
- Add comprehensive unit tests for all components
- Maintain full backward compatibility with existing flows

The implementation supports both IPv4 and IPv6 addresses and gracefully
handles legacy flows without source IP information.

coutinhop commented:
/sem-approve

andrenth commented:
@matthewdupre source ports would be just for completeness but if it's infeasible, the source IP is the information that's usually needed.

ravikalla committed:
Add destination IP field to Goldmane flow logs

This commit adds the missing destination IP field to the Goldmane
flow logging system, ensuring complete IP address tracking for both
source and destination endpoints.

Changes:
Add dest_ip field (field 18) to FlowKey protobuf definition
Update FlowKeyDestination struct to include DestIP field
Add DestIP() getter method to FlowKey
Update all proto conversion functions to handle destination IP
Fix ConvertGoldmaneToFlowlog to properly convert both IPs
Update ConvertFlowlogToGoldmane to include destination IP
Add comprehensive test coverage for IPv4, IPv6, and edge cases

This fixes the issue where destination IPs were lost during the
FlowLog to Goldmane conversion process.

matthewdupre commented on goldmane/pkg/types/flow.go:
```diff
@@ -29,12 +29,15 @@ type FlowKeySource struct {
 	SourceName      string
 	SourceNamespace string
 	SourceType      proto.EndpointType
+	SourceIP        string
```
Adding new fields to the Key has a fairly big impact on scalability.  Even just IP address will increase the potential number of records needed very significantly - for example if a large Deployment is connecting to another large Deployment right now that will only create two Flows (one for egress and one for ingress), but with IP addresses that would increase to n (or perhaps n^2 if each pod talks to many pods).

I suggest instead adding arrays of IPs to the Flow; so the number of Flows remains the same and we don't need to duplicate all the other data for each different source IP address.

We could also place a configurable limit on the number of IPs included, which will cap the total amount of data (not just the number of flows).  When Goldmane needs to combine multiple Flows with the same FlowKey it can then merge the arrays and truncate to the limit.

matthewdupre reviewed (changes requested):
I have one scalability related concern with the API change I've described below.  Please let me know if you need any further explanation or help making the adjustment.

ravikalla committed:
Move source IP/port from FlowKey to Flow arrays for scalability

Addresses reviewer feedback to prevent exponential FlowKey growth.

matthewdupre commented on goldmane/proto/api.proto:
```diff
@@ -334,6 +334,15 @@ message FlowKey {
   EndpointType source_type = 3;
 
+  // SourceIP is the IP address of the source endpoint.
+  string source_ip = 16;
```
This shouldn't be here: source_ips lives in the Flow

matthewdupre commented on goldmane/proto/api.proto:
```diff
@@ -334,6 +334,15 @@ message FlowKey {
+  string source_ip = 16;
+
+  // SourcePort is the source port number.
+  int64 source_port = 17;
```
As mentioned above, please remove source ports for now; let's get IPs working and then worry about ports if we get a compelling use case

matthewdupre commented on goldmane/proto/api.proto:
```diff
@@ -334,6 +334,15 @@ message FlowKey {
+  int64 source_port = 17;
+
+  // DestIP is the IP address of the destination endpoint.
+  string dest_ip = 18;
```
Please move this to the Flow and make it repeated

matthewdupre reviewed (commented):
Few leftover bits of API change

keliansb commented:
Not stale
